uproot/behaviors/TGraphAsymmErrors.py

```diff
@@ -0,0 +1,48 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/uproot4/blob/main/LICENSE
+
+"""
+This module defines the behaviors of ``TGraphAsymmErrors``, not including ``TGraph``.
+"""
+
+from __future__ import absolute_import
+
+import numpy
+
+import uproot
+
+# dict_keys(['@fUniqueID', '@fBits', 'fName', 'fTitle', 'fLineColor', 'fLineStyle', 'fLineWidth', 'fFillColor', 'fFillStyle', 'fMarkerColor', 'fMarkerStyle', 'fMarkerSize', 'fNpoints', 'fX', 'fY', 'fFunctions', 'fHistogram', 'fMinimum', 'fMaximum', 'fEXlow', 'fEXhigh', 'fEYlow', 'fEYhigh'])
+
+class TGraphAsymmErrors(object):
+    """
+    Behaviors for TGraphAsymmErrors, not including ``TGraph``
+    """
+
+    def values(self):
```

Comment:
I like the names `values` and `errors` (as methods, not properties), since this matches TH*.

I just deleted a lot of review because I'm thinking that they shouldn't be returning tuples that users need to unpack. For instance, if someone really wants just the low error bars on _y_, they'd have to say things like

```python
hist.errors()[1][0]
```

I know that `self.member("fEYlow")` is also an option, but most people won't discover that when there's a documented `errors` method available.

The TH* behaviors have an `axis` keyword that works like NumPy's (`axis=0` and `axis=-2` are the _x_ axis; `axis=1` and `axis=-1` are the _y_ axis) with ROOT-like extensions (`axis="x"` and `axis="y"`). If it also had a `dir` argument, each thing could be extracted without having to know the output tuple format:

```python
hist.values("x")
hist.values("y")
hist.errors("x", "low")
hist.errors("x", "high")
hist.errors("y", "low")
hist.errors("y", "high")
```

and then we could get this packed format by introducing a meaning for `"both"` on `axis` and `dir`:

```python
hist.values("both")
hist.errors("both", "both")
```

Part of the issue with returning both is that I would have guessed that the interface returns objects of shape `(N, 2)`, rather than `(2, N)`. That is, you have it returning the equivalent of `numpy.vstack([_x, _y])` (and `vstack` is more vectorized than iterating over `_x` and `_y` in `numpy.asarray([_x, _y])`), whereas I would have expected `numpy.dstack([_x, _y])[0]`:

```
>>> numpy.vstack([_x, _y])
array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14]], dtype=int32)
>>> numpy.dstack([_x, _y])[0]
array([[ 0,  5],
       [ 1,  6],
       [ 2,  7],
       [ 3,  8],
       [ 4,  9],
       [ 5, 10],
       [ 6, 11],
       [ 7, 12],
       [ 8, 13],
       [ 9, 14]], dtype=int32)
```

I think of the first axis as being the one ranging over a large dataset. (That's how Pandas would index it, too.) There are some NumPy interfaces that return a _tuple_ of 2 arrays, but not an array of shape `(2, N)`. So I just think the `(N, 2)` form (`dstack`) or 2-tuple (no "stack") would be more expected.

With `errors(axis="both", dir="both")`, should it be a 3D vector, in which the second and third dimension both have length 2? Or a 2D vector, in which the second dimension has length 4? (I think more structure is better, so 3D.) However it's set up, there's going to be some guessing or searching documentation to determine whether the two axes are paired more closely than the two directions or vice-versa (though most people would guess that _x_ comes before _y_ and _low_ comes before _high_).

Getting six 1D arrays is more typing, but it's the only unambiguous case. As a user, I'd want to have that ability, so that I don't have to look up documentation.  `:)`

Reply:
Thanks for the heavily detailed review here. I've implemented everything in the latest set of changes. I do agree that we need to try and reduce any amount of ambiguity and the best way to do this is to return (in as few calls as makes sense) simple 1-dimensional arrays where possible.